src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1578,45 +1605,126 @@ public Text getUsage(CommandSource src) {
     }
 
     /**
-     * Creates a new {@link CommandElement} with an {@link Entity} target
-     * and the provided {@link Text} key.
+     * Expect an argument to represent an {@link Entity}.
      *
-     * @param key The text key representation
-     * @return The command element
+     * @param key The key to store under
+     * @return the argument
      */
     public static CommandElement entity(Text key) {
-        return new EntityCommandElement(key, false);
+        return new EntityCommandElement(key, false, false, null);
     }
 
     /**
-     * Creates a new {@link CommandElement} with an {@link Entity} target
-     * or other source and the provided {@link Text} key.
+     * Expect an argument to represent an {@link Entity}, or if the argument is
+     * not present and the {@link CommandSource} is an entity, return the
+     * source.
      *
-     * @param key The text key representation
-     * @return The command element
+     * @param key The key to store under
+     * @return the argument
      */
     public static CommandElement entityOrSource(Text key) {
-        return new EntityCommandElement(key, true);
+        return new EntityCommandElement(key, true, false, null);
+    }
+
+    /**
+     * Expect an argument to represent an {@link Entity} of the specified type.
+     *
+     * @param key The key to store under
+     * @param type The type which the entity must subclass
+     * @return the argument
+     */
+    public static CommandElement entity(Text key, Class<? extends Entity> type) {
+        return new EntityCommandElement(key, false, false, type);
+    }
+
+    /**
+     * Expect an argument to represent an {@link Entity} of the specified
+     * {@link EntityType}.
+     *
+     * @param key The key to store under
+     * @param type The type which the entity must be
+     * @return the argument
+     */
+    public static CommandElement entity(Text key, EntityType type) {
+        return new EntityCommandElement(key, false, false, type.getEntityClass());
```

Comment:
This might not work as expected. Passing `ZOMBIE` will include `PIG_ZOMBIE` as well since `EntityPigZombie` extends `EntityZombie`.

Reply:
TIL. I'll get right on that.